controllers/medusa/medusabackupjob_controller.go

```diff
@@ -0,0 +1,290 @@
+				} else {
+					logger.Error(err, "backup failed", "CassandraPod", pod.Name)
+				}
+				backupMutex.Lock()
+				defer backupMutex.Unlock()
+				defer wg.Done()
+				backup.Status.InProgress = utils.RemoveValue(backup.Status.InProgress, pod.Name)
+				if succeeded {
+					backup.Status.Finished = append(backup.Status.Finished, pod.Name)
+				} else {
+					backup.Status.Failed = append(backup.Status.Failed, pod.Name)
+				}
+			}()
+		}
+		wg.Wait()
+		logger.Info("finished backup operations")
+		if err := r.Status().Patch(context.Background(), backup, patch); err != nil {
+			logger.Error(err, "failed to patch status", "Backup", fmt.Sprintf("%s/%s", backup.Name, backup.Namespace))
+		}
+	}()
+
+	return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+}
+
+func (r *MedusaBackupJobReconciler) getCassandraDatacenterPods(ctx context.Context, cassdc *cassdcapi.CassandraDatacenter, logger logr.Logger) ([]corev1.Pod, error) {
+	podList := &corev1.PodList{}
+	labels := client.MatchingLabels{cassdcapi.DatacenterLabel: cassdc.Name}
+	if err := r.List(ctx, podList, labels); err != nil {
+		logger.Error(err, "failed to get pods for cassandradatacenter", "CassandraDatacenter", cassdc.Name)
+		return nil, err
+	}
+
+	pods := make([]corev1.Pod, 0)
+	pods = append(pods, podList.Items...)
+
+	return pods, nil
+}
+
+func (r *MedusaBackupJobReconciler) createMedusaBackup(ctx context.Context, backup *medusav1alpha1.MedusaBackupJob, logger logr.Logger) error {
+	// Create a prepare_restore medusa task to create the mapping files in each pod.
+	// Returns true if the reconcile needs to be requeued, false otherwise.
+	logger.Info("Creating MedusaBackup object", "MedusaBackup", backup.ObjectMeta.Name)
+	// Create backups that should exist but are missing
+	backupKey := types.NamespacedName{Namespace: backup.ObjectMeta.Namespace, Name: backup.ObjectMeta.Name}
+	backupResource := &medusav1alpha1.MedusaBackup{}
+	if err := r.Get(ctx, backupKey, backupResource); err != nil {
+		if errors.IsNotFound(err) {
+			// Backup doesn't exist, create it
+			startTime := backup.Status.StartTime
+			finishTime := metav1.Now()
+			backupResource = &medusav1alpha1.MedusaBackup{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      backup.ObjectMeta.Name,
+					Namespace: backup.ObjectMeta.Namespace,
+				},
+				Spec: medusav1alpha1.MedusaBackupSpec{
+					CassandraDatacenter: backup.Spec.CassandraDatacenter,
+					Type:                backup.Spec.Type,
+				},
+			}
+			if err := r.Create(ctx, backupResource); err != nil {
+				logger.Error(err, "failed to create backup", "MedusaBackup", backup.ObjectMeta.Name)
+				return err
+			} else {
+				logger.Info("Created Medusa Backup", "Backup", backupResource)
+				// Read the backup again. Multiple attempts may be necessary due to caches, which is why the Get operation is in a loop.
+				backupInstance := &medusav1alpha1.MedusaBackup{}
+				found := false
+				for stay, timeout := true, time.After(10*time.Second); stay; {
```

Comment:
Blocking like this in a controller is an anti-pattern. By default each controller is configured to use a single goroutine. More importantly the loop is unnecessary. You can simply apply the patch after creating the object.

Reply:
very right, I forgot I had created a loop there as well.
I'll remove it.